u_http.go

```diff
@@ -0,0 +1,565 @@
+	https uhttpCloseableTransport
+
+	// h2 is like https but for the "h2" ALPN.
+	h2 uhttpCloseableTransport
+
+	// connCache maps a specific dialing address to an open
+	// connection. We will store open connections created by the
+	// dialOnly in this field. The https and h2 transports
+	// will get their new connections from this field. We will
+	// initialize this field during the first RoundTrip call.
+	connCache map[string][]net.Conn
+
+	// hostCache maps a specific URL.Host[:port] to the proper
+	// HTTP transport. We will remember which URL.Host[:port] wants
+	// HTTP/1.1 and which one wants H2. We will initialize
+	// this field during the first RoundTrip call.
+	hostCache map[string]http.RoundTripper
+
+	// mu allows for synchronized access of internals.
+	mu sync.Mutex
+}
+
+// _ ensures that UHTTPTransport matches the http.RoundTripper interface.
+var _ http.RoundTripper = &UHTTPTransport{
+	Proxy: http.ProxyFromEnvironment,
+}
+
+// UHTTPDefaultTransport is the default UHTTPTransport.
+var UHTTPDefaultTransport http.RoundTripper = &UHTTPTransport{}
+
+// errUHTTPNoCachedConn indicates that there are no cached connections.
+var errUHTTPNoCachedConn = errors.New("no cached conn")
+
+// errUHTTPUseH2 indicates that we should be using h2.
+var errUHTTPUseH2 = errors.New("utls: use h2")
+
+// errUHTTPUseHTTPS indicates that we should be using http/1.1 over TLS.
+var errUHTTPUseHTTPS = errors.New("utls: use https")
+
+// uhttpProxyURLKey is the type key to bind a proxy URL to a context.
+type uhttpProxyURLKey struct{}
+
+// uhttpWithProxyURL returns a copy of the current context
+// that keeps track of the current proxy URL. If there
+// is no proxy URL, this function returns the original context.
+func uhttpWithProxyURL(ctx context.Context, proxyURL *url.URL) context.Context {
+	if proxyURL == nil {
+		return ctx
+	}
+	return context.WithValue(ctx, uhttpProxyURLKey{}, proxyURL)
+}
+
+// uhttpContextWithProxyURL returns the proxy URL that
+// we saved in the context so we can honor Proxy. If the
+// user configured no proxy, then we return nil.
+func uhttpContextWithProxyURL(ctx context.Context) *url.URL {
+	URL, _ := ctx.Value(uhttpProxyURLKey{}).(*url.URL)
+	return URL
+}
+
+// RoundTrip implements http.RoundTripper.RoundTrip.
+func (txp *UHTTPTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	txp.maybeInitTxps()
+	// Step 1: immediately dispatch HTTP requests
+	switch req.URL.Scheme {
+	case "http":
+		uhttpLog.Printf("uhttp: using transport %s", txp.cleartext)
+		return txp.cleartext.RoundTrip(req)
+	case "https":
+		// we need to figure out which transport to use - fallthrough
+	default:
+		return nil, errors.New("uhttp: unsupported URL scheme")
+	}
+	// Step 2: check whether we have a proxy URL.
+	proxyURL, err := txp.proxy(req)
+	if err != nil {
+		return nil, err
+	}
+	// Step 3: dispatch HTTPS requests to the proper transport
+	child := txp.hostCacheGetOrDefault(req.URL)
+	const maxRetries = 4
```

Comment:
We should allow users to change maxRetries, perhaps as a field in UHTTPTransport